example-plugins/sethome.lua

```diff
@@ -0,0 +1,94 @@
+local lukkit = require("lukkit")
+
+local homestore_opts = {
+  store_type: lukkit.store.type.JSON,
+  overwrite_oncorrupt: false
+}
+
+
+-- Runs when the plugin is enabled
+function lukkit.plugin:enable()
+  local plugin_name = self.info:get_name();
+  self.logger:info("Enabling {}", plugin_name)
+end
+
+-- Runs when the plugin is disabled (shutdown, reload etc.)
+function lukkit.plugin:disable()
+  local plugin_name = self.info:get_name();
+  self.logger:info("Disabling {}", plugin_name)
+end
+
+-- Runs when the plugin is loaded (before enabled, on file read, can stop loading)
+function lukkit.plugin:load(load_context)
+  -- Can load a SQLite DB, JSON doc or YAML
+  local home_storage = self.store:load("home", homestore_opts)
+
+  -- do setup with store
+
+  if (home_storage:is_read_only()) then
+    -- Stop the plugin from continuining with a load
+    load_context:reject("Home store cannot be written to")
+  end
+end
+
+
+function lukkit.plugin:command(cmd, sender, args)
```

Comment:
DISC: not using function definition for commands.

Reply:
I would rather prefer to have a listener based approach than this, because:
1. That would allow me to separate commands into files and each "command" could setup its own listener
2. Lukkit doesn't have to rely on users defining functions with proper known names (ex. typo in function name would go uncaught)

Reply:
I'm in two minds about this:

I'm totally for 'modularising' commands into files/handlers. That's *ideal* and the way I would like to write commands.

My issue with this is that Lukkit's target audience. The project leans towards people newer to programming/plugin development who might not understand those concepts that easily. We could do both which would cater to both parties. Just trying to think as if I was new to programming and what would be the easiest for me to pick up.

Reply:
Well, I think it would be totally doable to have a more complex base that is very flexible and a little bit harder to grasp AND a simpler abstraction built on top of that. When I was using the lukkit v2 implementation I was already abstracting away the lukkit interfaces away so my code would be easier to maintain.

> Little bit of background: I have never written "production" Lua before interacting with Lukkit and only toyed a little bit with it in Love engine and some other embedded solutions, but I have a lot of experience in other languages.

One thing that I immediately started doing is abstracting away from using Lukkit as a single plugin for one thing. Instead, I wanted Lukkit to be a platform to many smaller plugins that allow me to achieve things in a way that I want them to be. I split up all my microplugins into separate files, created a Plugin class with lifecycle methods baked in and a plugin registry that relayed lukkit events into my microplugins. At the same time, I didn't want the code to be in separate folders as recommended through the docs because I've created a bunch of helpers that I would share between microplugins.

I imagine this may be similiar to what other people try to achieve when using Lukkit, because once they create 1 singular plugin thing using Lukkit to solve a particular problem, it feels natural to continue to add new functionality that maybe unrelated to the initial problem. I think it would be really important to have that story reflected in the API in a way that encourages clean and maintainable code. On the other hand I understand that some people may ever only want to run one thing to solve their problem and thats it. 

I believe ditching yaml as a configuration format for the plugins is a way to go and approaching this from the perspective of "code as a platform" where all plugins are created dynamically. This allows to loading new plugins and unloading old plugins in realtime without the need to restart the lukkit master plugin itself.

Reply:
Would you mind sharing an example (or *the* example you mentioned) so I can understand this a bit more? I get where you're coming from but seeing how someone has approached it would be more valuable for me. 

As for YAML config - I have a disdain for it and want it gone too. I've considered two options:
 1. Allow configuration using a single Lua file (`config.lua`) where it takes the config builder via a `require` and goes from there. This gets called before the rest of the plugin is loaded in to satisfy the CraftBukkit plugin manager.
 2. Configure the plugin at runtime. Much more flexible but the downside of this method is we lose that integration with Bukkit's plugin manager.

In v2 someone suggested integrating with the native functionality of the CraftBukkit plugin manager. It gives Lukkit plugins a more natural feel with support for `/help`, `/plugins` and other bits. If config was moved into the runtime we'd lose that and I'm not sure how I feel about it.

Reply:
> Configure the plugin at runtime. Much more flexible but the downside of this method is we lose that integration with Bukkit's plugin manager.
Not necessarily, we can execute the Lua code up to the point before it's enabled to get the plugin config. Isnt that what you'd have to do with a `config.lua` file anyways? We could also use reflections or something like that to override the default plugin manager

Reply:
Using reflections causes two issues:
 - relies on modifying `private` data structures which could change at any point
 - reduces compatibility between CraftBukkit versions or custom Bukkit implementations

If you do runtime-based configuration you need to manage those things internally instead of passing them off to the server. That's why I'm iffy about doing that.

Just to clarify: by runtime I mean Lua plugin execution and not server runtime.

Reply:
Hmm, I see the issue with it now. A `config.lua` does seem like the best idea then.